File: python/1238. Circular Permutation in Binary Representation.py
class Solution(object):
    def circularPermutation(self, n, start):
        """
        :type n: int
        :type start: int
        :rtype: List[int]
        """
        return [start, start ^ 1] if n == 1 else self.circularPermutation(n - 1, start) + self.circularPermutation(n - 1, start ^ 1 << n - 1)[::-1] 

File: go/src/leetcode/1238/1238. Circular Permutation in Binary Representation.go
package main

/**
https://leetcode.cn/problems/circular-permutation-in-binary-representation/description/
思路：格雷编码的生成过程, G(i) = i ^ (i/2);
time: O(1<<n)
space: O(1<<n)
*/
func circularPermutation(n int, start int) []int {
	var ans = make([]int, 1<<n)
	for i := range ans {
		ans[i] = i ^ i>>1
		ans[i] = ans[i] ^ start
	}
	return ans
}

